api/agent/pure_runner.go

```diff
@@ -559,9 +583,28 @@ func (pr *pureRunner) AddCallListener(cl fnext.CallListener) {
 	pr.a.AddCallListener(cl)
 }
 
+func (pr *pureRunner) saveCallHandle(ch *callHandle) {
+	pr.callHandleLock.Lock()
+	pr.callHandleMap[ch.c.Model().ID] = ch
+	pr.callHandleLock.Unlock()
+}
+
+func (pr *pureRunner) removeCallHandle(cID string) {
+	pr.callHandleLock.Lock()
+	delete(pr.callHandleMap, cID)
+	pr.callHandleLock.Unlock()
+
+}
+
 func (pr *pureRunner) spawnSubmit(state *callHandle) {
 	go func() {
+		isAcksync := state.c.Type == models.TypeAcksync
+		// we keep track of the callHandle which will be used in the process to send the ack back
+		if isAcksync {
+			pr.saveCallHandle(state)
+		}
 		err := pr.a.Submit(state.c)
+
```

Comment:
we might as well call removeCallHandle() here too, but really not important. :-)